### PR TITLE
Snapshot Lisk Core - Closes #49

### LINKED ISSUE #49: Snapshot Lisk Core
### Description

The migrator tool after being able to successfully connect to the Lisk Core node should be able to observe the current chain height and issue a snapshot command to capture the Lisk Core node’s state and generate the snapshot file.

### Acceptance Criteria

- Lisk Migrator is able to observe the chain height on the node
- As soon as the chain height reaches the snapshot height (`snapshot-height` flag), cache the block ID and issue the command to snapshot the state on the Lisk Core node
- After the state snapshot is created successfully, observe the chain for the block at the snapshot height to be finalized
- Verify the block ID at the snapshot height is exactly the same after finalization
- Stop the Lisk Core v3 node
- Create a new DB instance to be used for the migration process using `lisk-db`
- After the block at snapshot height is finalized, apply the generated snapshot to the migration DB instance using `lisk-db`
- Start the snapshot block building process

### PR DESCRIPTION
**Pull request by nagdahimanshu**
### What was the problem?

This PR resolves #49 

### How was it solved?

- [x] Lisk Migrator is able to observe the chain height on the node
- [x] As soon as the chain height reaches the snapshot height (snapshot-height flag), cache the block ID and issue the command to snapshot the state on the Lisk Core node
- [x] After the state snapshot is created successfully, observe the chain for the block at the snapshot height to be finalized
- [x] Verify the block ID at the snapshot height is exactly the same after finalization
- [x] Stop the Lisk Core v3 node
- [x] Create a new DB instance to be used for the migration process using lisk-db
- [x] After the block at snapshot height is finalized, apply the generated snapshot to the migration DB instance using lisk-db
- [x] Start the snapshot block building process

### How was it tested?
Local

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -86,25 +94,24 @@ class LiskMigrator extends Command {
...
-			char: 'w',
-			required: true,
-			env: 'SNAPSHOT_WAIT_THRESHOLD',
+		'snapshot-path': flagsParser.string({
+			char: 'p',
+			required: false,
 			description:
-				'Blocks to wait before creating a snapshot. Applies only if NODE_ENV=test otherwise 201 value be used.',
-			default: 201,
+				'Path where the state snapshot will be creating. Current directory will be considered the default if not provided.',
```
Comment (sameersubudhi) [inline src/index.ts:101]:
```suggestion
				'Path where the state snapshot will be created. When not supplied, defaults to the current directory.',
```

File: src/index.ts [hunk 2]
```diff
@@ -148,44 +155,50 @@ class LiskMigrator extends Command {
...
 			height: snapshotHeight,
 			delay: 500,
+			isFinal: false,
 		});
 
+		await setBlockIDAtSnapshotHeight(liskCorePath, snapshotHeight);
+
+		// TODO: Placeholder to issue createSnapshot command from lisk-core
+		cli.action.start('Creating snapshot');
+		await createSnapshot(snapshotPath);
```
Comment (sameersubudhi) [inline src/index.ts:165]: This should also take the `liskCorePath` as input.

File: src/index.ts [hunk 3]
```diff
@@ -148,44 +155,50 @@ class LiskMigrator extends Command {
...
+
+		const blockID = getBlockIDAtSnapshotHeight();
+		const finalizedBlockID = await getBlockIDAtSnapshotHeightFinalized(
+			liskCorePath,
+			snapshotHeight,
+		);
+
+		cli.action.start('Verifying blockID');
+		if (blockID !== finalizedBlockID) {
+			this.error('BlockID doesnt match with the finalized blockID');
```
Comment (sameersubudhi) [inline src/index.ts:184]:
```suggestion
			this.error('Snapshotted blockID doesn't match with the finalized blockID.');
```

File: src/index.ts [hunk 4]
```diff
@@ -148,44 +155,50 @@ class LiskMigrator extends Command {
...
+		}
+		cli.action.stop();
+
+		// TODO: Stop lisk core automatically if stop command is available from SDK
```
Comment (sameersubudhi) [inline src/index.ts:188]:
```suggestion
		// TODO: Stop lisk core automatically when the application management is implemented
```

File: src/utils/chain.ts [hunk 5]
```diff
@@ -15,18 +15,43 @@
 import cli from 'cli-ux';
 import { getClient } from '../client';
 
+let blockID: string;
```
Comment (sameersubudhi) [inline src/utils/chain.ts:18]: Please rename the variable to be more intuitive.
```suggestion
let blockIDAtSnapshotHeight: string;
```

File: src/utils/chain.ts [hunk 6]
```diff
@@ -15,18 +15,43 @@
...
 	readonly label: string;
 	readonly height: number;
 	readonly liskCorePath: string;
 	readonly delay: number;
+	readonly isFinal: boolean;
 }
 
-export const getChainHeight = async (liskCorePath: string): Promise<number> => {
+export const getChainHeight = async (liskCorePath: string, isFinal: boolean): Promise<number> => {
 	const client = await getClient(liskCorePath);
```
Comment (sameersubudhi) [inline src/utils/chain.ts:29]: Please consider renaming this method to `getAPIClient`. Also, please consider creating two separate methods: `getCurrentFinalizedHeight` and `getCurrentChainHeight`.
```suggestion
	const client = await getAPIClient(liskCorePath);
```

File: src/utils/chain.ts [hunk 7]
```diff
@@ -15,18 +15,43 @@
...
+	liskCorePath: string,
+	height: number,
+): Promise<string> => {
+	const client = await getClient(liskCorePath);
+	const result = await client.block.getByHeight(height);
+	blockID = result.header.id.toString('hex');
+	return blockID;
+};
+
+export const getBlockIDAtSnapshotHeight = (): string => blockID;
```
Comment (sameersubudhi) [inline src/utils/chain.ts:45]: A setter should not return any value. 
```suggestion
export const setBlockIDAtSnapshotHeight = async (
	liskCorePath: string,
	height: number,
): void => {
	const client = await getClient(liskCorePath);
	const result = await client.block.getByHeight(height);
	blockIDAtSnapshotHeight = result.header.id.toString('hex');
};

export const getBlockIDAtSnapshotHeight = (): string => blockIDAtSnapshotHeight;
```

File: src/utils/chain.ts [hunk 8]
```diff
@@ -15,18 +15,43 @@
...
+
+export const getBlockIDAtSnapshotHeightFinalized = async (
```
Comment (sameersubudhi) [inline src/utils/chain.ts:47]: This method can be named generic.
```suggestion
export const getBlockIDAtHeight = async (
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]
Please consider the following.

File: src/client.ts [hunk 9]
```diff
@@ -15,7 +15,7 @@ import { createIPCClient } from '@liskhq/lisk-api-client';
 
 let client: object;
 
-export const getClient = async (liskCorePath: string): Promise<any> => {
+export const getAPIClient = async (liskCorePath: string): Promise<any> => {
```
Comment (sameersubudhi) [inline src/client.ts:18]:
```suggestion
export const getAPIClient = async (liskCorePath: string): Promise<APIClient> => {
```

File: src/utils/chain.ts [hunk 10]
```diff
@@ -13,22 +13,48 @@
...
+	blockIDAtSnapshotHeight = result.header.id.toString('hex');
+};
+
+export const getBlockIDAtSnapshotHeight = (): string => blockIDAtSnapshotHeight;
+
+export const getBlockIDAtHeight = async (liskCorePath: string, height: number): Promise<string> => {
+	const client = await getAPIClient(liskCorePath);
+	const result = await client.block.getByHeight(height);
+	const blockIDAtSnapshotHeightFinalized = result.header.id.toString('hex');
+	return blockIDAtSnapshotHeightFinalized;
```
Comment (sameersubudhi) [inline src/utils/chain.ts:55]:
```suggestion
	const blockID = result.header.id.toString('hex');
	return blockID;
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: src/utils/chain.ts [hunk 11]
```diff
@@ -13,22 +13,48 @@
  */
 
 import cli from 'cli-ux';
-import { getClient } from '../client';
+import { getAPIClient } from '../client';
```
Comment (sameersubudhi) [inline src/utils/chain.ts:16]: Please run `npm i -S @liskhq/lisk-chain` and make the following changes to replace `Record<string, any>` with `Record<string, unknown>` below.

```suggestion
import { Block, BlockHeader } from '@liskhq/lisk-chain';
import { getAPIClient } from '../client';
```

File: src/utils/chain.ts [hunk 12]
```diff
@@ -13,22 +13,48 @@
...
+	return result.finalizedHeight;
+};
+
+export const setBlockIDAtSnapshotHeight = async (
+	liskCorePath: string,
+	height: number,
+): Promise<void> => {
+	const client = await getAPIClient(liskCorePath);
+	const result: Record<string, any> = await client.block.getByHeight(height);
+	blockIDAtSnapshotHeight = result.header.id.toString('hex');
```
Comment (sameersubudhi) [inline src/utils/chain.ts:46]:
```suggestion
	const result = await client.block.getByHeight(height) as Record<string, Block>;
	const blockHeader = result.header as unknown as BlockHeader;
	blockIDAtSnapshotHeight = blockHeader.id.toString('hex');
```

File: src/utils/chain.ts [hunk 13]
```diff
@@ -13,22 +13,48 @@
...
+};
+
+export const getBlockIDAtSnapshotHeight = (): string => blockIDAtSnapshotHeight;
+
+export const getBlockIDAtHeight = async (liskCorePath: string, height: number): Promise<string> => {
+	const client = await getAPIClient(liskCorePath);
+	const result: Record<string, any> = await client.block.getByHeight(height);
+	const blockID = result.header.id.toString('hex');
```
Comment (sameersubudhi) [inline src/utils/chain.ts:54]:
```suggestion
	const result: Record<string, unknown> = await client.block.getByHeight(height);
	const blockHeader = result.header as unknown as BlockHeader;
	const blockID = blockHeader.id.toString('hex');
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: src/types.ts [hunk 14]
```diff
@@ -42,3 +42,47 @@ export interface UnregisteredAccount {
 export interface UnregisteredAddresses {
 	readonly unregisteredAddresses: UnregisteredAccount[];
 }
+
+export interface Transaction {
```
[the remark this answers is not in the document]
Comment (nagdahimanshu) [reply src/types.ts]: Yes, i used from SDK but forget to remove. Thanks

**Review by ManuGowda — APPROVED** [review]
👍🏼

**Pull request merged**